### PR TITLE
refactor: Update `pyproject.toml`, fix lints, typing

### LINKED ISSUE #9: Keep `pyproject.toml` in sync with `altair`
I think generally this would be desired, but enough of a low-priority item to have been overlooked.

# Motivation
## Reduce docs build time
@binste I did some [investigating](https://github.com/vega/altair/pull/3431/commits/ca961c067506f4a914d1aff3884668489fc57b31) on improving the performance when building the `altair` docs a while back.
I ultimately [reverted](https://github.com/vega/altair/pull/3431/commits/26df622644227d9e309444ad021d8dd2c964eabe) those changes, as that alone had no effect on build time.

`sphinxext_altair` is quite small, but used frequently enough that I think it (if optimized) could speed things up a bit.
At the very least, ruling it out would be helpful, as would any lessons learned (w/ `sphinx`, `docutils`, `jinja2`) during the process when moving back to look at [sphinxext](https://github.com/vega/altair/tree/main/sphinxext).

Applying `ruff` auto-fixes would be an easy starting point for this work

# `pyproject.toml`
## Links
- [`altair`](https://github.com/vega/altair/blob/cfab6601aa649227c20880d05ae27799297694c3/pyproject.toml)
- [`sphinxext_altair`](https://github.com/vega/sphinxext-altair/blob/b937d72f9cb392254380bd529d6d6daa1daac859/pyproject.toml)

## Differences
- Probably won't benefit from *all* of the current `ruff` rules, but would be simple enough to copy over `[tool.ruff.(lint|format)]`
- Usage of `black`
	- Shouldn't be required since `ruff` added a formatter
- python version bounds (also internally inconsistent)
	- https://github.com/vega/sphinxext-altair/blob/b937d72f9cb392254380bd529d6d6daa1daac859/pyproject.toml#L31
	- https://github.com/vega/sphinxext-altair/blob/b937d72f9cb392254380bd529d6d6daa1daac859/pyproject.toml#L104
	- https://github.com/vega/sphinxext-altair/blob/b937d72f9cb392254380bd529d6d6daa1daac859/.github/workflows/lint.yml#L14
	- https://github.com/vega/sphinxext-altair/blob/b937d72f9cb392254380bd529d6d6daa1daac859/pyproject.toml#L91
	- https://github.com/vega/sphinxext-altair/blob/b937d72f9cb392254380bd529d6d6daa1daac859/.github/workflows/build.yml#L10

### PR DESCRIPTION
**Pull request by dangotbanned**
Resolves https://github.com/vega/sphinxext-altair/issues/9

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -7,12 +7,12 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
Comment (dangotbanned) [inline .github/workflows/build.yml]: Version bounds now match `altair`

File: pyproject.toml [hunk 2]
```diff
@@ -37,11 +36,11 @@ classifiers= [
     "License :: OSI Approved :: BSD License",
     "Natural Language :: English",
     "Framework :: Sphinx :: Extension",
-    "Programming Language :: Python :: 3.7",
     "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
```
Comment (dangotbanned) [inline pyproject.toml:42]: Declaring 3.12 support, removing 3.7

File: tests/test_altairplot.py [hunk 3]
```diff
@@ -90,7 +93,7 @@ def test_altairplotdirective(app):
   // embed when document is loaded, to ensure vega library is available
   // this works on all modern browsers, except IE8 and older
   document.addEventListener("DOMContentLoaded", function(event) {
-      var spec = {"config": {"view": {"continuousWidth": 300, "continuousHeight": 300}}, "data": {"values": [{"x": "A", "y": 5}, {"x": "B", "y": 3}, {"x": "C", "y": 6}, {"x": "D", "y": 7}, {"x": "E", "y": 2}]}, "mark": {"type": "bar"}, "encoding": {"x": {"field": "x", "type": "nominal"}, "y": {"field": "y", "type": "quantitative"}}, "$schema": "https://vega.github.io/schema/vega-lite/v5.8.0.json"};
+      var spec = {"config": {"view": {"continuousWidth": 300, "continuousHeight": 300}}, "data": {"values": [{"x": "A", "y": 5}, {"x": "B", "y": 3}, {"x": "C", "y": 6}, {"x": "D", "y": 7}, {"x": "E", "y": 2}]}, "mark": {"type": "bar"}, "encoding": {"x": {"field": "x", "type": "nominal"}, "y": {"field": "y", "type": "quantitative"}}, "$schema": "https://vega.github.io/schema/vega-lite/v5.20.1.json"};
```
Comment (dangotbanned) [inline tests/test_altairplot.py:96]: Ideally these wouldn't be set manually, but fixing that can happen in another PR.

**Comment by dangotbanned** [comment]
@binste I'm not sure how clear the commit trail is, but all of the changes I've made in `altairplot.py` are one or more of:

- minor performance increase
- fixing lints
- fixing typing
- reducing repetition

There should be no functional differences (assuming the current test suite is sufficient)

**Review by binste — APPROVED** [review]
Thanks, LGTM! Didn't spot any issues with the changes, commits were clear, and it's great to have the code bases more aligned :) This repo hasn't gotten much attention since we took it out of the Altair repo.

I leave the PR open in case you want to drop Python 3.8 support at the same time. up to you.

**Comment by dangotbanned** [comment]
> Thanks, LGTM! Didn't spot any issues with the changes, commits were clear, and it's great to have the code bases more aligned :) This repo hasn't gotten much attention since we took it out of the Altair repo.
> 
> I leave the PR open in case you want to drop Python 3.8 support at the same time. up to you.

Thanks @binste 

~~I think bumping the `3.8` refs will trigger a lot of additional `ruff` violations/fixes.
I'm happy to handle that in another PR since this is passing everything for now?~~

# Edit
Only one fix, which was a version guard I added earlier in this PR.

**Review by binste — APPROVED** [review]
LGTM, thanks again!

**Pull request merged**